servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyIoExecutors.java

```diff
@@ -39,35 +40,89 @@ private NettyIoExecutors() {
+     * {@link IoExecutor} threads.
      */
+    @Deprecated
     public static EventLoopAwareNettyIoExecutor createIoExecutor(ThreadFactory threadFactory) {
         return createIoExecutor(getRuntime().availableProcessors() * 2, threadFactory);
     }
 
     /**
      * Create a new {@link NettyIoExecutor}.
      *
+     * @param <T> Type of the IO thread instances created by factory.
      * @param ioThreads number of threads.
-     * @param threadFactory the {@link ThreadFactory} to use.
+     * @param threadFactory the {@link ThreadFactory} to use. If possible you should use an instance of
+     * {@link NettyIoThreadFactory} as it allows internal optimizations.
+     * @return The created {@link IoExecutor}
+     */
+    public static <T extends Thread & IoThreadFactory.IoThread> EventLoopAwareNettyIoExecutor createIoExecutor(
+            int ioThreads, IoThreadFactory<T> threadFactory) {
+        validateIoThreads(ioThreads);
+        return new EventLoopGroupIoExecutor(createEventLoopGroup(ioThreads, threadFactory), true);
+    }
+
+    /**
+     * Create a new {@link NettyIoExecutor}.
+     *
+     * @param ioThreads number of threads.
+     * @param threadFactory the {@link ThreadFactory} to use. If possible you should use an instance of
+     * {@link NettyIoThreadFactory} as it allows internal optimizations.
      * @return The created {@link IoExecutor}
+     * @deprecated Future versions of ServiceTalk will require a {@link IoThreadFactory} for creating
+     * {@link IoExecutor} threads.
      */
+    @Deprecated
     public static EventLoopAwareNettyIoExecutor createIoExecutor(int ioThreads, ThreadFactory threadFactory) {
         validateIoThreads(ioThreads);
         return new EventLoopGroupIoExecutor(createEventLoopGroup(ioThreads, threadFactory), true);
     }
 
+    /**
+     * Create a new {@link EventLoopGroup}.
+     *
+     * @param <T> Type of the IO thread instances created by factory.
+     * @param ioThreads number of threads
+     * @param threadFactory the {@link ThreadFactory} to use.
+     * @return The created {@link IoExecutor}
+     */
+    public static <T extends Thread & IoThreadFactory.IoThread> EventLoopGroup createEventLoopGroup(int ioThreads,
+            IoThreadFactory<T> threadFactory) {
+        validateIoThreads(ioThreads);
+        return isEpollAvailable() ? new EpollEventLoopGroup(ioThreads, threadFactory) :
+                isKQueueAvailable() ? new KQueueEventLoopGroup(ioThreads, threadFactory) :
+                        new NioEventLoopGroup(ioThreads, threadFactory);
+    }
+
     /**
      * Create a new {@link EventLoopGroup}.
      *
      * @param ioThreads number of threads
      * @param threadFactory the {@link ThreadFactory} to use.
      * @return The created {@link IoExecutor}
+     * @deprecated Future versions of ServiceTalk will require a {@link IoThreadFactory} for creating
+     * {@link IoExecutor} threads.
      */
+    @Deprecated
```

Comment:
Looking at `fromNettyEventLoop` and `fromNettyEventLoopGroup`. This is unfortunate for the use-case of detecting `IoThread` as we don't have control how that `EventLoop[Group]` is created. 
This is internal API, but some rare users potentially can use those methods. They are useful when you already have a netty-based application and you would like to share `EventLoop` between existing app/library and ST.
@bondolo any ideas how to make a predicate for conditional offloading operators more reliable?

Reply:
Not really unfortunately. Other than the general `IoThread` mechanism we haven't come up with another good approach for identifying cases where we should offload. Solutions which identify which threads belong to which executors are complicated and would be annoying to retrofit back in to existing `ThreadFactory` and `Executor` implementations. The most universal conditional offloading predicate would be to ask the target `Executor` if the current thread is one of theirs and if it returns false, then offload. The current approach for #1724 of using the IoThread marker interface is not as complete but sufficient for protecting event loop threads.